File: facilitator-solutions/javascript/rung-6-zero-means-prime.js
// Rung 6 - stop refilling the array. Includes rungs 1 to 5.
//
// A new Uint32Array is already zeroed. Invert the sense so that ZERO means prime
// and the fill disappears entirely - a whole pass over the array, deleted, every
// lap. The runtime was already zeroing it; we stopped fighting that.
//
// THE TAIL WORD BUG. The array holds ceil(count / 32) words, so the last word has
// bits for numbers past n. Those bits are zero, and zero now means prime. At
// n = 1,000,000 that is 16 spare bits covering 1,000,001 to 1,000,031, of which
// 1,000,003 is prime - so the sieve reports 78,499 primes and fails validation.
//
// Bound results by count, never by words.length * 32. This only bites once rungs 5
// and 6 are both present: rung 5 alone fills with ones, and a Uint8Array has no
// padding to get wrong.

export const getSieve = (sieveSize) => {
	const count = (sieveSize + 1) >> 1;
	const words = new Uint32Array((count + 31) >>> 5);

	const limit = Math.floor(Math.sqrt(sieveSize));

	return {
		run: () => {
			for (let factor = 3; factor <= limit; factor += 2) {
				const f = factor >> 1;

				if ((words[f >>> 5] & (1 << f)) === 0) {
					for (let q = (factor * factor) >> 1; q < count; q += factor) {
						words[q >>> 5] |= 1 << q;
					}
				}
			}
		},
		results: () => {
			const primes = [];
			if (sieveSize >= 2) {
				primes.push(2);
			}
			// Bounded by count, not words.length * 32 - see the tail word note.
			for (let i = 1; i < count; i++) {
				if ((words[i >>> 5] & (1 << i)) === 0) {
					primes.push(2 * i + 1);
				}
			}
			return primes;
		},
	};
};

File: facilitator-solutions/csharp/rung-6-zero-means-prime.cs
using System;
using System.Collections.Generic;

namespace Race;

// Rung 6 - stop refilling the array. Includes rungs 1 to 5.
//
// new ulong[n] hands back zeroed memory. Invert the sense so that ZERO means prime
// and the fill disappears entirely - a whole pass over the array, deleted, every
// lap. The runtime was already zeroing it; we just stopped fighting that.
//
//     test prime    (words[i >> 6] & (1UL << i)) == 0
//     mark composite words[i >> 6] |= 1UL << i
//
// THE TAIL WORD BUG. The word array covers ceil(_count / 64) words, so the last
// word holds bits for numbers beyond n. Those bits are zero, and zero now means
// prime. At n = 1,000,000 that is 32 spare bits covering 1,000,001 to 1,000,063,
// of which 1,000,003 is prime - so the sieve reports 78,499 primes and fails.
//
// The fix is to bound Result by _count and never by _words.Length * 64. Note this
// only bites once rungs 5 and 6 are BOTH present: rung 5 alone fills with ones, and
// a byte array has no padding to get wrong.
public class Sieve
{
	private const int WordBits = 64;

	private readonly int _sieveSize;
	private readonly int _count;
	private readonly ulong[] _words;

	public Sieve(int sieveSize)
	{
		_sieveSize = sieveSize;
		_count = (sieveSize + 1) / 2;
		_words = new ulong[(_count + WordBits - 1) / WordBits];
	}

	public void Run()
	{
		int limit = (int)Math.Sqrt(_sieveSize);

		for (int factor = 3; factor <= limit; factor += 2)
		{
			int f = factor >> 1;

			if ((_words[f >> 6] & (1UL << f)) == 0)
			{
				for (int q = (factor * factor) >> 1; q < _count; q += factor)
				{
					_words[q >> 6] |= 1UL << q;
				}
			}
		}
	}

	public List<int> Result()
	{
		List<int> primes = [];

		if (_sieveSize >= 2)
		{
			primes.Add(2);
		}

		// Bounded by _count, not by _words.Length * 64 - see the tail word note above.
		for (int i = 1; i < _count; i++)
		{
			if ((_words[i >> 6] & (1UL << i)) == 0)
			{
				primes.Add(2 * i + 1);
			}
		}

		return primes;
	}
}
